### PR TITLE
Add integration tests using a proxy with mTLS for control plane with Elastic Defend installed

### PR DESCRIPTION
**Pull request by AndersonQ**
## What does this PR do?

Add integration tests for the Elastic Agent running Elastic Defend with a mTLS proxy
See https://github.com/elastic/elastic-agent/issues/5716 for tje covered test cases

## Why is it important?

To have automated tests covering mTLS and Elastic Defend

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- ~~[ ] I have made corresponding changes to the documentation~~
- ~~[ ] I have made corresponding change to the default configuration files~~
- [x] I have added tests that prove my fix is effective or that my feature works
- ~~[ ] I have added an entry in `./changelog/fragments` using the [changelog tool]~~(https://github.com/elastic/elastic-agent#changelog)
- [x] I have added an integration test or an E2E test

## Disruptive User Impact

 - N/A

## How to test this PR locally

In short, manually reproduce the tests described on https://github.com/elastic/elastic-agent/issues/5716.

The long version:
 - run 3 proxies:
   - mTLScli -> a proxy with mTLS to be passed as cli arguments
   - mTLSpolicy -> a proxy with mTLS configured through the policy
   - oneWayTLSpolicy -> a proxy with simple/one way TLS which defines its CA in the policy
   - all the certificates must be RSA because of an Elastic Defend restriction

 - set up the proxies:
   - add a test to `testing/integration/endpoint_security_test.go` to configure and run the needed proxies:
```go
func TestDebugTestProxies(t *testing.T) {
	prepareProxies(t, &url.URL{Host: "a.wrong.fleet.host"}, "the.real.fleet.host")

	t.Logf("CTRL + C to exit")
	<-make(chan struct{})
}
```
 
   - run the test with:
```
go test -v -timeout 0 -tags integration -run TestDebugTestProxies ./testing/integration/

=== RUN   TestDebugTestProxies
    endpoint_security_test.go:1333: [mtlsCLI] certificates saved on: /tmp/TestDebugTestProxies-mtlsCLI2475114678
    endpoint_security_test.go:1333: [mtlsPolicy] certificates saved on: /tmp/TestDebugTestProxies-mtlsPolicy1317235685
    endpoint_security_test.go:1333: [oneWayTLSPolicy] certificates saved on: /tmp/TestDebugTestProxies-oneWayTLSPolicy327431283
```

   - check the logs like `[%s] certificates saved on: %s`. they will show where the certificates are.
   - set up the necessary fleet hosts and proxies on fleetUI
   - install/enroll the agent using the configured fleet hosts and proxies. Again, check the test cases on https://github.com/elastic/elastic-agent/issues/5716

## Related issues
- Closes https://github.com/elastic/elastic-agent/issues/5491
- Closes https://github.com/elastic/elastic-agent/issues/5716

## Questions to ask yourself

- How are we going to support this in production?
- How are we going to measure its adoption?
- How are we going to debug this?
- What are the metrics I should take care of?
- ...

### CODE REVIEW

**Comment by AndersonQ** [comment]
folks (@blakerouse, @pchila, @kaanyalti), all the tests are passing now. In case you were waiting the tests to pass to review it

**Review by blakerouse — COMMENTED** [review]
Overall this looks good.

I do question the change from returning errors to using `require.NoError` directly. Why was that change made?

**Comment by AndersonQ** [comment]
> I do question the change from returning errors to using `require.NoError` directly. Why was that change made?

to follow the convention we have on our tests. The majority of the tests use this approach

File: testing/integration/endpoint_security_test.go [hunk 1]
```diff
@@ -894,3 +901,536 @@ func TestForceInstallOverProtectedPolicy(t *testing.T) {
...
+	fleetKibanaHostmTLS, err := stack.KibanaClient.CreateFleetServerHosts(ctx, kibana.ListFleetServerHostsRequest{
+		ID:        "proxyPolicymTLS" + testUUID,
+		Name:      "proxyPolicymTLS" + testUUID,
+		HostURLs:  []string{fleethostWrong.String()},
+		IsDefault: false,
+		ProxyID:   fleetProxymTLS.Item.ID,
+	})
+	require.NoError(t, err, "error creating fleet host with mTLS proxy")
+
+	// add new fleet-server host with oen way TLS proxy
```
Comment (pchila) [inline testing/integration/endpoint_security_test.go:961]:
```suggestion
	// add new fleet-server host with one way TLS proxy
```

File: testing/integration/endpoint_security_test.go [hunk 2]
```diff
@@ -894,3 +901,536 @@ func TestForceInstallOverProtectedPolicy(t *testing.T) {
...
+					buff.Reset()
+
+					got, err := f.ExecInspect(ctx)
+					if err != nil {
+						buff.WriteString(fmt.Sprintf("error running inspect cmd: %v", err))
+						return false
+					}
+
+					return proxyPolicymTLS.URL == got.Fleet.ProxyURL
+				}, time.Minute, time.Second, "inspect never showed proxy from policy: %s", buff)
```
Comment (pchila) [inline testing/integration/endpoint_security_test.go:1150]: `buff` is really the last error encountered in the assert.Eventually, maybe we can define and treat it as such:



```suggestion
				// feel free to change the variable name
				var buff error
				assert.Eventuallyf(t, func() bool {

					got, err := f.ExecInspect(ctx)
					if err != nil {
						buff = fmt.Errorf("error running inspect cmd: %w", err)
						return false
					}
					// (optional) reset the error
					buff = nil
					return proxyPolicymTLS.URL == got.Fleet.ProxyURL
				}, time.Minute, time.Second, "inspect never showed proxy from policy: %s", buff)
```

File: testing/integration/endpoint_security_test.go [hunk 3]
```diff
@@ -894,3 +901,536 @@ func TestForceInstallOverProtectedPolicy(t *testing.T) {
...
+	defaultFleetHost := fleetServerURL[8:]
+	fleethostWrong, err := url.Parse("https://fixme.elastic.co:443")
+	require.NoError(t, err, "failed parsing fleethostWrong")
+
+	// ================================= proxy =================================
+	mtlsCLI, mtlsPolicy, oneWayTLSPolicy, proxyCLI, proxyPolicymTLS, proxyPolicyOneWayTLS := prepareProxies(t, fleethostWrong, defaultFleetHost)
+
+	// =================== Prepare fleet hosts and proxies ===================
+
+	// create mTLS proxy
```
Comment (pchila) [inline testing/integration/endpoint_security_test.go:931]: Is it possible to create the policies and enrollment tokens as part of each testcase instead of preparing a bunch of different policies and tokens upfront and only have the testcases specify the token use for enrollment? I know that there's a 1-1 relationship between the enrollment token and the policy but this setup makes the test kinda hard to follow:
- the policy definition is far removed from the assertions on the agent behavior
- each test case enroll exactly once with a given token and the a big assertion function is called to wait for the policy to apply and check the results
- it's kinda difficult to follow which test uses which policy (I know, I know, enrollment token... but still...)

File: testing/integration/endpoint_test_tools.go [hunk 4]
```diff
@@ -128,6 +128,6 @@ func installElasticDefendPackage(t *testing.T, info *define.Info, policyID strin
 		t.Logf("Error installing fleet package: %v", err)
 		return r, fmt.Errorf("error installing fleet package: %w", err)
 	}
-	t.Logf("Endpoint package Policy Response:\n%+v", pkgResp)
```
Comment (pchila) [inline testing/integration/endpoint_test_tools.go:131]: was this log not useful for debugging when tests fail ?

**Review by pchila — COMMENTED** [review]
A few comments here and there, mostly about readability

File: testing/integration/endpoint_security_test.go [hunk 5]
```diff
@@ -894,3 +901,536 @@ func TestForceInstallOverProtectedPolicy(t *testing.T) {
 	out, err := fixture.Exec(ctx, args)
 	require.Errorf(t, err, "No error detected, command output: %s", out)
 }
+
+func TestInstallDefendWithMTLSandEncCertKey(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (AndersonQ) [reply testing/integration/endpoint_security_test.go:905]: what about mTLS -> `TestInstallDefendWithmTLSAndEncCertKey`?

File: testing/integration/endpoint_security_test.go (hunk 3 above)
Comment (AndersonQ) [reply testing/integration/endpoint_security_test.go:931]: I did that way to minimise the code for each test case and let it be reusable.
Just the `enrollmentTokenOneWayProxyResp` is used only once. the others are used at least by 2 tests.

I get your point, but I believe it'd be worse having each test to set it up. It'd be more repetitive as each test would need to set it up and the test cases would be a lot longer, making it harder to see what each test do.

File: testing/integration/endpoint_test_tools.go (hunk 4 above)
Comment (AndersonQ) [reply testing/integration/endpoint_test_tools.go:131]: the response is returned regardless, even on failure. So it's better to log the response if and only if an error happens. Here it was being logged on success as well.

**Review by pchila — APPROVED** [review]
LGTM
